src/Utilities/DereferenceWrapper.hpp

```diff
@@ -10,6 +10,7 @@
 #include <utility>
```

Comment:
I'm struggling to follow this commit. It seems to be doing several things at once. Is it necessary to do all of them? There is so much happening in the tests it makes it quite difficult to follow. If it has to happen all at once I'll keep going but if it doesn't it would make my life a lot easier to not need to check 10 different changes at once :)

Edit:
After looking through more of this I think the best way to go forward is to just pull this commit into a separate PR, split the commit to get each as focused as possible (if it's not possible that's fine :) ) and then it should be fairly straightforward to review by the looks of things. Let's hold off on the documentation commit and do that separately at the end once we have the others merged so the interface for things is clearly defined :)